Sources/SwiftProtobuf/TextFormatEncoder.swift

```diff
@@ -33,57 +33,91 @@ private let tab = [UInt8](repeating: asciiSpace, count: tabSize)
 
 /// TextFormatEncoder has no public members.
 internal struct TextFormatEncoder {
+    private static let foyerSize = 4096
+    private var foyer = Array<UInt8>(repeating: 0, count: TextFormatEncoder.foyerSize)
+    private var foyerIndex = 0
+
     private var data = [UInt8]()
     private var indentString: [UInt8] = []
-    var stringResult: String {
-        get {
-            return String(bytes: data, encoding: String.Encoding.utf8)!
+
+    internal mutating func constructFinalResult() -> String {
+        if foyerIndex > 0 {
+            let subfoyer = foyer[0..<foyerIndex]
+            data.append(contentsOf: subfoyer)
+            foyerIndex = 0
+        }
+        return String(bytes: data, encoding: String.Encoding.utf8)!
+    }
+
+    // Append each byte to the foyer, append and
+    // recycle the foyer when it fills up.
+    private mutating func append(byte: UInt8) {
+        if foyerIndex > TextFormatEncoder.foyerSize - 1 {
+            data.append(contentsOf: foyer)
+            foyerIndex = 0
         }
+        foyer[foyerIndex] = byte
+        foyerIndex += 1
     }
 
     internal mutating func append(staticText: StaticString) {
         let buff = UnsafeBufferPointer(start: staticText.utf8Start, count: staticText.utf8CodeUnitCount)
-        data.append(contentsOf: buff)
+        for b in buff {
+            append(byte: b)
+        }
     }
 
     internal mutating func append(name: _NameMap.Name) {
-        data.append(contentsOf: name.utf8Buffer)
+        for b in name.utf8Buffer {
+            append(byte: b)
+        }
```

Comment:
This change looks suspect to me. `append(contentsOf:)` is already supposed to do the right thing and it avoids repeatedly checking capacity.

Reply:
This is sorta what I was referencing, this would have to atleast do the "flush" of the other buffer before it could directly append.